### PR TITLE
[mir-opt] SimplifyLocals should also clean up debuginfo

### PR DESCRIPTION
**Pull request by scottmcm**
When not asking for debuginfo or when only asking for line-info, there's no need to treat a `Local` as live just because it's referenced in a piece of debug info.

This is particularly true for inlined functions -- today <https://rust.godbolt.org/z/KqsrjfGja> we keep around a bunch of useless things like debug info for the parameter to `wrapping_neg` inside `wrapping_sub` inside `wrapping_byte_sub` inside `post_inc_start` inside `slice::Iter::next`, even though we never actually *use* that local because its reads/writes got optimized away.

That meant that a bunch of mir-opt tests had everything interesting optimize away with this change, so there are a bunch of changes in here to properly mark them as unit tests or to pass `-C debuginfo=2` as a way to suppress removing a bunch of this dead code.

### CODE REVIEW

File: tests/mir-opt/inline/unwrap_unchecked.unwrap_unchecked.PreCodegen.after.mir [hunk 1]
```diff
@@ -5,8 +5,7 @@ fn unwrap_unchecked(_1: Option<T>) -> T {
     let mut _0: T;                       // return place in scope 0 at $DIR/unwrap_unchecked.rs:+0:54: +0:55
     scope 1 (inlined #[track_caller] Option::<T>::unwrap_unchecked) { // at $DIR/unwrap_unchecked.rs:10:9: 10:27
         debug self => _1;                // in scope 1 at $SRC_DIR/core/src/option.rs:LL:COL
-        let mut _2: &std::option::Option<T>; // in scope 1 at $SRC_DIR/core/src/option.rs:LL:COL
```
Comment (scottmcm) [inline tests/mir-opt/inline/unwrap_unchecked.unwrap_unchecked.PreCodegen.after.mir:8]: This file is a nice demonstration: the function had a stray `&Option` that got inlined from `unwrap_unchecked` that was never calculated, but still had a `StorageLive`+`StorageDead` because it was the `self` to `Option::is_some`.

Now it's removed as irrelevant, as seems entirely reasonable to me.

**Comment by scottmcm** [comment]
libcore.rlib: 0.3% smaller (56,292,736 → 56,090,538)
liballoc.rlib: 1.7% smaller (7,029,814 → 6,909,206)

**Comment by cjgillot** [comment]
I love the idea of removing useless cruft from MIR, but I fear touching debuginfo could backfire.

I we were to land this PR, we'll probably get a another report akin to https://github.com/rust-lang/rust/issues/103655.
We currently have 2 PRs to fix that issue with `DeadStoreElimination`, blocked on a decision on the behaviour of mir-opt levels.

The harder part is that debuginfo is not crate-local: we may need to preserve more debuginfo in MIR in case the downstream crate doing the codegen needs it.

Removing storage statements on never-borrowed locals in `RemoveStorageMarkers` is probably the low-hanging fruit here.

**Comment by scottmcm** [comment]
@cjgillot Do we really need to preserve per-variable debug info MIR no matter how the current crate is being compiled?  That seems unfortunate.  Why isn't it ok to say that if you want to debug into a dependency's variables, it needs to be compiled with debuginfo ≥ 2?  After all, this change *isn't* removing variable debug info -- even when it's strictly useless -- if the session is asking for per-variable debug info.

If that's not ok, what would you think of a middle-ground: Today writes to a local never count as a use, but I could instead change it so that writes to a local count as a use *iff* that local is in debuginfo.  That would allow removing the ones that are entirely useless -- there's no point in having debuginfo point at a local if that local is only ever `undef` -- but would keep from optimizing away writes to unused-except-for-debuginfo locals.

cc @wesleywiser, as the author of the linked bug report.

**Comment by Nilstrieb** [comment]
> The harder part is that debuginfo is not crate-local: we may need to preserve more debuginfo in MIR in case the downstream crate doing the codegen needs it.

I don't think this makes sense. Semantically, monomorphizations are still part of the upstream crate and should follow its opts. If I compile regex without debuginfo because I want maximal performance but do enable debuginfo for my own code, I would not expect my regex instances to get debuginfo (worsening their performance).

File: compiler/rustc_mir_transform/src/simplify.rs [hunk 2]
```diff
@@ -632,6 +650,16 @@ impl<'tcx> Visitor<'tcx> for UsedLocals {
             self.use_count[local] -= 1;
         }
     }
+
+    fn visit_var_debug_info(&mut self, var_debug_info: &VarDebugInfo<'tcx>) {
```
Comment (scottmcm) [inline compiler/rustc_mir_transform/src/simplify.rs:654]: Another idea here: I think we have a spare byte in `VarDebugInfo`, so we could put the [`DebugInfo`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_session/config/enum.DebugInfo.html) level on each of these.

That would let us be aware of the original debuginfo level from the original code even when it's inlined, rather than depending on the level from the current crate.

**Comment by JakobDegen** [comment]
Yeah, my one reaction here is that while I agree that respecting the downstream `debuginfo` is silly, if we are going to stop doing that we should at least do the correct thing in the other direction - ie if we compile a dep with `-Cdebuginfo=2`, that's the amount of debuginfo it should get, even if it's codegen'd downstream. I don't know how hard to implement this is.

**Comment by cjgillot** [comment]
I convinced myself that this PR, as written, is the simpler and more consistent solution. Since there hasn't been a clear statement of policy on the matter of opt-level/debuginfo interaction, this deserves a FCP.

# Change summary

This PR starts using the `-Cdebuginfo` flag to allow MIR optimizations to degrade debuginfo in MIR. This is a change from the current policy in which MIR opts preserve debuginfo at the expense of optimizations. I believe it consistent with the discussion in https://rust-lang.zulipchat.com/#narrow/stream/238009-t-compiler.2Fmeetings/topic/.5Bsteering.5D.202023-03-17.20scope.20and.20goals.20of.20rust-lang.2Frust.20opti in which using two flags instead of one (opt-level and debuginfo) was discussed.

In the current behaviour, MIR always have full debug info whatever `-Cdebuginfo` says, and the crate doing codegen decides what is the amount to give to LLVM.

In the proposed behaviour, MIR will respect the debuginfo level of the crate that MIR-optimizes it. This means that generics from a dependency compiled with `-Cdebuginfo=0` will have degraded debuginfo, whatever the flags for the downstream crates are. The downstream crate doing codegen will still decide what is gives to LLVM, but may have less to start with.

This change is user visible, although the behaviour of `-Cdebuginfo` on MIR is currently undocumented. The impact is however unlikely, except:
- manually invoked rustc with different debuginfo levels;
- `rlib` files shipped by a vendor with low debuginfo level.

This creates many different combinations of debuginfo, weird cases with MIR inlining. However, I think this is the right approach: if the user requests a specific debuginfo level on one crate, possibly different from the one of other crates, we should respect this request, and leverage it into optimizations.

@rfcbot merge

**Comment by JakobDegen** [comment]
I think it's worth mentioning that the opt level policy stuff which I've been discussing was not meant to cover this case - I agree that generic functions not obeying the debug info level of the crate they are defined in is basically just a bug.

File: tests/mir-opt/pre-codegen/mem_replace.mem_replace.PreCodegen.after.mir [hunk 3]
```diff
@@ -38,15 +34,11 @@ fn mem_replace(_1: &mut u32, _2: u32) -> u32 {
...
         _3 = &raw const (*_1);           // scope 2 at $SRC_DIR/core/src/mem/mod.rs:LL:COL
-        StorageLive(_5);                 // scope 2 at $SRC_DIR/core/src/mem/mod.rs:LL:COL
         _0 = (*_3);                      // scope 5 at $SRC_DIR/core/src/ptr/mod.rs:LL:COL
-        StorageDead(_5);                 // scope 2 at $SRC_DIR/core/src/mem/mod.rs:LL:COL
         StorageDead(_3);                 // scope 2 at $SRC_DIR/core/src/mem/mod.rs:LL:COL
         StorageLive(_4);                 // scope 3 at $SRC_DIR/core/src/mem/mod.rs:LL:COL
         _4 = &raw mut (*_1);             // scope 3 at $SRC_DIR/core/src/mem/mod.rs:LL:COL
-        StorageLive(_6);                 // scope 3 at $SRC_DIR/core/src/mem/mod.rs:LL:COL
         (*_4) = _2;                      // scope 8 at $SRC_DIR/core/src/ptr/mod.rs:LL:COL
-        StorageDead(_6);                 // scope 3 at $SRC_DIR/core/src/mem/mod.rs:LL:COL
```
Comment (scottmcm) [inline tests/mir-opt/pre-codegen/mem_replace.mem_replace.PreCodegen.after.mir:49]: Rebased to pick up this test, as mentioned in https://github.com/rust-lang/rust/pull/111010/files#r1181176608

**Comment by wesleywiser** [comment]
Just to weigh in from my perspective, I think this change makes sense and I'm in favor. If we are not going to generate per variable debuginfo, we don't need to be tracking it at the MIR level. 

For code that is instantiated in downstream crates, I agree with @Nilstrieb's point that semantically the code should behave as if it was compiled as part of the originating crate with its compilation options. That's not how it works today but I don't think this PR changes that situation very significantly. In practice, most users seem to compile all the crates in their dependency graph with the same debuginfo level, so from what I've seen, this inconsistency is also not something many users will notice.

**Comment by scottmcm** [comment]
Hello @Aaron1011 @estebank @michaelwoerister @nagisa @nikomatsakis @pnkfelix!  Friendly 1-week ping for FCP [above](https://github.com/rust-lang/rust/pull/110702#issuecomment-1528821027).

**Comment by nikomatsakis** [comment]
@rfcbot fcp reviewed

Hmm. I am going to check my box but I also want to raise a mild concern (with a lower-case 'c') -- the experience of using Rust in a debugger is already somewhere between 'unusable' and 'not great', and from what I understand a lot of that has to do with LLVM not preserving debuginfo very well when doing optimizations. I agree that the behavior here makes sense but I am nervous about it giving us license to land an optimization and disable debuginfo rather than going the extra mile to land the optimization *and* find a way to preserve debuginfo.

In short, I think this policy makes sense, but I think that just as we go the extra mile to preserve span information and make great error messages, we should go the extra mile to preserve debug info.

File: compiler/rustc_mir_transform/src/simplify.rs [hunk 4]
```diff
@@ -645,6 +673,22 @@ fn remove_unused_definitions_helper(used_locals: &mut UsedLocals, body: &mut Bod
...
+            body.var_debug_info.retain(|info| {
+                let keep = match info.value {
+                    VarDebugInfoContents::Place(place) => used_locals.is_used(place.local),
+                    _ => true,
+                };
+
+                if !keep {
+                    trace!("removing var_debug_info {:?}", info);
+                    modified = true;
+                }
```
Comment (cjgillot) [inline compiler/rustc_mir_transform/src/simplify.rs:686]: Changing `modified` here is not useful: we don't update `used_locals`, so we do not mark new locals as unused.

**Comment by cjgillot** [comment]
r=me once FCP finishes

File: tests/mir-opt/pre-codegen/slice_index.slice_get_unchecked_mut_range.PreCodegen.after.mir [hunk 5]
```diff
@@ -19,11 +19,7 @@ fn slice_get_unchecked_mut_range(_1: &mut [u32], _2: std::ops::Range<usize>) ->
                 let mut _8: usize;       // in scope 3 at $SRC_DIR/core/src/slice/index.rs:LL:COL
                 let mut _9: *mut u32;    // in scope 3 at $SRC_DIR/core/src/slice/index.rs:LL:COL
                 let mut _10: usize;      // in scope 3 at $SRC_DIR/core/src/slice/index.rs:LL:COL
-                let _16: std::ops::Range<usize>; // in scope 3 at $SRC_DIR/core/src/slice/index.rs:LL:COL
-                let mut _17: std::ops::Range<usize>; // in scope 3 at $SRC_DIR/core/src/intrinsics.rs:LL:COL
-                let mut _18: *mut [u32]; // in scope 3 at $SRC_DIR/core/src/intrinsics.rs:LL:COL
```
Comment (scottmcm) [inline tests/mir-opt/pre-codegen/slice_index.slice_get_unchecked_mut_range.PreCodegen.after.mir:24]: Oh, fun, #111813 moved these useless fields to the end, so this diff is really clean after the rebase.

**Comment by scottmcm** [comment]
@bors r-

Failure repro'd; not sure how this PR is causing BBs to be missing.  Will try to figure it out.

File: compiler/rustc_mir_transform/src/simplify.rs (hunk 2 above)
Comment (cjgillot) [reply compiler/rustc_mir_transform/src/simplify.rs:654]: Now that we have SROA in MIR, we may need to check `VarDebugInfoContents::Composite` too.

File: compiler/rustc_mir_transform/src/simplify.rs [hunk 6]
```diff
@@ -632,6 +650,16 @@ impl<'tcx> Visitor<'tcx> for UsedLocals {
             self.use_count[local] -= 1;
         }
     }
+
+    fn visit_var_debug_info(&mut self, var_debug_info: &VarDebugInfo<'tcx>) {
+        if let VarDebugInfoContents::Place(..) = var_debug_info.value
```
Comment (cjgillot) [inline compiler/rustc_mir_transform/src/simplify.rs:655]: Now that we have SROA in MIR, we may need to check `VarDebugInfoContents::Composite` too.

Comment (scottmcm) [reply compiler/rustc_mir_transform/src/simplify.rs]: For now I'm counting any debuginfo that's not just a direct mention of a local as a use of the variable that keeps it from being removed -- so my expectation is that I don't need to handle `Composite`, as it'll always end up calling `super_var_debug_info`, which will add a use for each local.

Any guesses at what might be wrong here, @cjgillot?  It's still failing in mipsel but [working](https://github.com/rust-lang/rust/actions/runs/5363521273/jobs/9731157734) in the normal  build, so I'm very puzzled.

I saw you mention https://github.com/rust-lang/rust/pull/109025#discussion_r1239491410 that you're waiting on this, but I'm unsure how to get it into a good place right now.

**Comment by scottmcm** [comment]
Let's see if this build also hits the failure:
@bors try

**Comment by Dylan-DPC** [comment]
@scottmcm any updates on this?

**Comment by Dylan-DPC** [comment]
Closing this as inactive. Feel free to reöpen this pr or create a new pr if you get the time to work on this. Thanks

File: compiler/rustc_mir_transform/src/simplify.rs [hunk 7]
```diff
@@ -581,3 +628,20 @@ impl<'tcx> MutVisitor<'tcx> for LocalUpdater<'tcx> {
...
+    }
+
+    match opts.debuginfo {
+        DebugInfo::None | DebugInfo::LineDirectivesOnly | DebugInfo::LineTablesOnly => false,
+        DebugInfo::Limited | DebugInfo::Full => true,
+    }
+}
+
+// For now we only remove basic debuginfo, like `foo => _3`, and don't attempt
+// to clean up more complicated things like `foo => Foo { .0 => _2, .1 => _4 }`
```
Comment (cjgillot) [inline compiler/rustc_mir_transform/src/simplify.rs:644]: Why not `foo => _3.projection`?
Why not composite fields?

File: compiler/rustc_mir_transform/src/simplify.rs [hunk 8]
```diff
@@ -640,8 +642,23 @@ fn preserve_debug_even_if_never_generated(opts: &Options) -> bool {
...
+                Ok(opt @ None) => *opt = Some(local),
+                Ok(Some(current)) if *current == local => {}
+                res @ Ok(Some(_)) => *res = Err(()),
+            }
+        }
+    }
+
+    let mut finder = SingleLocalFinder(Ok(None));
+    finder.visit_var_debug_info(info);
+    if let Ok(Some(local)) = finder.0 { Some(local) } else { None }
```
Comment (cjgillot) [inline compiler/rustc_mir_transform/src/simplify.rs:663]: `Index` projections are forbidden in debuginfo, so this can be simplified.

File: compiler/rustc_mir_transform/src/simplify.rs [hunk 9]
```diff
@@ -547,7 +547,10 @@ impl<'tcx> Visitor<'tcx> for UsedLocals {
     }
 
     fn visit_var_debug_info(&mut self, var_debug_info: &VarDebugInfo<'tcx>) {
-        if !self.preserve_debug && debug_info_is_for_simple_local(var_debug_info).is_some() {
+        // We don't want to have to track *conditional* uses (such as
+        // "`_4` is used iff `_5` is used" from `debug x => Foo(_4, _5)`),
+        // so if this mentions multiple locals we just treat them all as used.
```
Comment (cjgillot) [inline compiler/rustc_mir_transform/src/simplify.rs:552]: Should we just drop parts of the composite? If `_4` is unused, go from `Foo { .0 => _4, .1 => _5 }` to just `Foo { .1 => _5 }`?
This line can definitely just be a `if !self.preserve_debug`, and complexity will go in `remove_unused_definitions_helper`.
